pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1205 @@ func (o *ClusterUninstaller) validate() error {
+		}
+		for _, arn := range arns {
+			if _, ok := deleted[arn]; !ok {
+				err = deleteARN(awsSession, arn, o.Logger)
+				if err == nil {
+					deleted[arn] = exists
+				} else {
+					err = errors.Wrapf(err, "deleting %s", arn)
+					o.Logger.Debug(err)
+					loopError = err
+				}
+			}
 		}
 	}
 
 	return nil
 }
 
-func deleteRunner(deleteFuncName string, dFunction deleteFunc, awsSession *session.Session, filters Filter, clusterName string, logger logrus.FieldLogger, channel chan string) {
-	backoffSettings := wait.Backoff{
-		Duration: time.Second * 10,
-		Factor:   1.3,
-		Steps:    100,
+func tagMatch(filters []Filter, tags map[string]string) bool {
+	for _, filter := range filters {
+		match := true
+		for filterKey, filterValue := range filter {
+			tagValue, ok := tags[filterKey]
+			if !ok {
+				match = false
+				break
+			}
+			if tagValue != filterValue {
+				match = false
+				break
+			}
+		}
+		if match {
+			return true
+		}
 	}
+	return len(filters) == 0
+}
 
-	err := wait.ExponentialBackoff(backoffSettings, func() (bool, error) {
-		return dFunction(awsSession, filters, clusterName, logger)
-	})
+type iamRoleSearch struct {
+	client    *iam.IAM
+	filters   []Filter
+	logger    logrus.FieldLogger
+	unmatched map[string]struct{}
+}
 
-	if err != nil {
-		logger.Fatalf("Unrecoverable error/timed out: %v", err)
-		os.Exit(1)
+func (search *iamRoleSearch) arns() ([]string, error) {
+	if search.unmatched == nil {
+		search.unmatched = map[string]struct{}{}
 	}
 
-	// record that the goroutine has run to completion
-	channel <- deleteFuncName
-	return
-}
+	arns := []string{}
+	var err error
+	err2 := search.client.ListRolesPages(
+		&iam.ListRolesInput{},
+		func(results *iam.ListRolesOutput, lastPage bool) bool {
+			for _, role := range results.Roles {
+				if _, ok := search.unmatched[*role.Arn]; ok {
+					continue
+				}
 
-func getAWSSession(region string) (*session.Session, error) {
-	awsConfig := &aws.Config{Region: aws.String(region)}
+				// Unfortunately role.Tags is empty from ListRoles, so we need to query each one
+				var response *iam.GetRoleOutput
+				response, err = search.client.GetRole(&iam.GetRoleInput{RoleName: role.RoleName})
+				if err != nil {
+					if err.(awserr.Error).Code() == iam.ErrCodeNoSuchEntityException {
+						search.unmatched[*role.Arn] = exists
+					} else {
+						err = errors.Wrapf(err, "get tags for %s", *role.Arn)
+						search.logger.Info(err)
+					}
+				} else {
+					role = response.Role
+					tags := make(map[string]string, len(role.Tags))
+					for _, tag := range role.Tags {
+						tags[*tag.Key] = *tag.Value
+					}
+					if tagMatch(search.filters, tags) {
+						arns = append(arns, *role.Arn)
+					} else {
+						search.unmatched[*role.Arn] = exists
+					}
+				}
+			}
+
+			return !lastPage
+		},
+	)
 
-	// Relying on appropriate AWS ENV vars (eg AWS_PROFILE, AWS_ACCESS_KEY_ID, etc)
-	s, err := session.NewSession(awsConfig)
 	if err != nil {
```

Comment:
should we fail on last failed `search.client.GetRole(&iam.GetRoleInput{RoleName: role.RoleName})` ?

Reply:
> should we fail on last failed...

These `err2` / `err` handlers after a `*Pages()` call give us the following logic:

* If something in the per-page handler errored out (setting `err`), return that error.  Otherwise fall back to:
* If something in the `*Pages()` call itself (setting `err2`), return that error.

For the handler-set `err`, I generally continue to do as much as I can with the rest of the paginated results, although we could reroll to return `false` in those cases to fail fast.  Did you want me to make any particular adjustment?

Reply:
> WDYT about replacing `err` with `lastErr` and `err2` with `err`

I've done this (using `lastError`) with b44da20 -> a6e65d3.  I've read through the diff between those commits a few times, and I don't *think* I've missed anything, but it touches a lot of lines, so it would be good to get another set of eyes on the updated code as well.

Reply:
> I've done this (using lastError)
`lastErr` would have been fine too ;) but this is good too.

> so it would be good to get another set of eyes on the updated code as well.
looking at https://github.com/openshift/installer/compare/b44da203195ef9e9433a0c08bf9eed6d8bb55276..a6e65d3155c0582954d0f1199fd440eea7cafd7e it seems okay